### ISSUE TITLE
several problems

### ISSUE THREAD

**Issue by bobpappas** [kushalmangtani/ayasdi-challenge#1]
This was an OK start, but did not solve several requirements:
- First of all, as of the last commit, the program doesn't complete and has an exception. I was able to fix it and move on, but you should fix that issue.
- Dataset produced is only 10,000, not 1 million rows
- Columns 2-10 do not have 10% NULLs
- Columns 11-19 are not using random words from the English Dictionary
- Columns 11-19 do not have 10% NULLs

I also would recommend a test to prove the resulting sqlite DB solves all the requirements regarding means, gaussian distribution, NULL count, and all dates in 2014.

**Comment by kushalmangtani** [comment]
1. I fixed the issue no 1. It does not throws an exception now. Must have been introduced when I was refactoring.
2.  So, it is producing 1000000 records in csv file, However adding only one chunk of size 10,000 and then the program exits. Will look at this.

```
Kushals-MBP:coding-challenge kushalmangtani$ wc -l ayasdi_assignment.csv
 1000001 ayasdi_assignment.csv
Kushals-MBP:coding-challenge kushalmangtani$ head ayasdi_assignment.csv
col1    col2_11 col3_12 col4_13 col5_14 col6_15 col7_16 col8_17 col9_18 col10_19        col11   col12   col13   col14   col15   col16   col17   col18   col19   col20
1       12.1998001689   13.974499787    15.4085233985   10.0909830183   8.847036224     25.787621366    21.5287614849   18.0419175716   16.3642694658   WOBC    AOFB    SmJp    gfjEknas        cxXI    occn    qMXg    LYhs    04/19/2014
2       11.5948777967   11.9007400675   14.9366151601   15.0260953038   10.8620394365   20.5709385258   14.3427030322   16.8518246625   24.6594318498   JGaw    RuZV    bPpG    HDrCyrDW        ncJm    rJDk    IGMm    LJWH    06/23/2014
3       9.05587342178   13.0939776646   9.98559374014   11.0451210878   10.0123411984   11.6210752601   14.9253326651   19.7204488294   14.6992889805   AFOO    gLtM    faiM    KDuSBjmG        PRQA    xxDy    JqYb    ftzz    03/22/2014
4       10.1956616074   12.0660410001   8.18073836113   13.5622328585   16.231595407    16.3320450412   16.4876789332   20.6304502726   24.788452595    okbE    jrGA    Vpqn    lwWdXXGq        jxpJ    Ahnu    vjYg    qOeG    09/07/2014
5       10.4280873298   12.1121951732   10.9422964335   16.4403047378   9.28360378868   7.64372191821   23.6823953912   15.7675489218   17.8911837644   EcVz    fTDl    YgBh    BHMtTMha        ZOnx    WWwA    dJOg    dVlC    08/27/2014
6       11.0030720396   10.008787489    12.0391755089   14.9791149342   20.6970808202   14.4050697225   6.5568424642    22.3071291507   24.723057625    FiQC    vHDj    qgYl    wjCFKYQn        oJuI    NZnz    Edlc    TVOk    11/23/2014
7       9.29069484204   12.5905761302   12.3781479541   17.9905084548   8.88480076348   19.3244537129   16.5529492934   26.51688971     26.502825152    HzLC    GvQc    nbGh    buUbUFPf        SuMq    meGR    pIMN    svnx    04/19/2014
8       9.85074111953   15.0962871108   9.6093176924    17.241160598    24.2979744845   19.8750528922   8.43351065889   21.1134680324   28.1469846665   wzRm    zIck    iGul    ofxqsmFN        SfLU    NNuu    padJ    VEVJ    04/09/2014
9       10.1927171842   10.6270241307   15.5648310271   20.6267579108   18.3366577245   13.9396282646   10.7853006461   35.0185823499   14.6808591844   hWjw    SIEZ    yAmt    vuZyfAOw        WPLJ    DtGj    JDET    UPby    12/04/2014
```

**Comment by kushalmangtani** [comment]
Ok, 2. is also fixed. Was not clearing the list after yielding a chunk.

```
2016-09-28 12:59:36,958 INFO => Inserting chunk no : 92. 
2016-09-28 12:59:37,214 INFO => Inserting chunk no : 93. 
2016-09-28 12:59:37,470 INFO => Inserting chunk no : 94. 
2016-09-28 12:59:37,766 INFO => Inserting chunk no : 95. 
2016-09-28 12:59:38,027 INFO => Inserting chunk no : 96. 
2016-09-28 12:59:38,295 INFO => Inserting chunk no : 97. 
2016-09-28 12:59:38,545 INFO => Inserting chunk no : 98. 
2016-09-28 12:59:38,795 INFO => Inserting chunk no : 99. 
2016-09-28 12:59:39,043 INFO => Inserting chunk no : 100. 
2016-09-28 12:59:39,212 INFO => Success => Inserted 1000000 entries in  ayasdi_assignment_table_kushalll
```

```
sqlite> select count(*) from ayasdi_assignment_table_kushalll;
1000000
```

Have checked in the changes.

**Comment by kushalmangtani** [comment]
3: Columns 11 - 19 are words generated using random letters taken from ascii.letters. Note: each word is of size 4. I can make the dynamic size by using a random size.
Is that the requirement ? 

4,5 : I was not sure how Col 2 - 10 and Col 11 -19 could generate nulls. The only thing I could think of was a weighted sum list like below:
list = [a] \* 45 + [b] \* 45 + [None] \* 10. 

This will generate a list with 100 values with 10 % null. Would this be ok ?

It will be helpful if you could provide me some examples.

**Comment by bobpappas** [comment]
The requirement is to use an English Dictionary to select the random words to put into those columns.

**Comment by bobpappas** [comment]
4,5: it is your responsibility to sprinkle NULLs in the columns at a 10% amount. This is similar to prepping a dataset in real life where there are NULLs in the data.

So for col11, there are NULLs in 10% of the entries. How you decide to implement that result is up to you.

**Comment by kushalmangtani** [comment]
Ok, I will spend some time on this tom. I have a couple of things in mind, i will try it.

**Comment by kushalmangtani** [comment]
I spent quite some time on this and achieved all the requirements(Notes below).
Also, I improved the performance significantly. The file takes approx 65 sec to now generate and the entire program is executed in less than 2 min.

Achieved 3: Created a dictionary of words and then invoked  random.choice on the dictionary keys.
Achieved 4,5 : Achieved this by using a mod operator. Wrote an empty string as a null entry in csv file and when writing to db, replaced the empty string with None. Was able to achieve approx 10 % NULL values.

Did not get enough time to write lot of tests. So, wrote some simple queries in the program itself and printed it out.
